### ISSUE TITLE
Feature Request: Extend Objects by id feature for Collections of Objects

### ISSUE THREAD

**Issue by panzerj** [vojtechhabarta/typescript-generator#316]
Thanks for implementing the jackson id feature so quickly. It works great - at least for single objects. Would it be possible to extend it for collections also?

Using my example from above:

Generates my id string properly:
@JsonIdentityReference(alwaysAsId = true) private ProjectComponent wheelType; 

Generates ProjectComponent[] instead an array of ids
@JsonIdentityReference(alwaysAsId = true) private List<ProjectComponent> wheelTypes;

_Originally posted by @panzerj in https://github.com/vojtechhabarta/typescript-generator/issues/298#issuecomment-470860470_

**Comment by vojtechhabarta** [comment]
I will check it.

**Comment by vojtechhabarta** [comment]
@panzerj I implemented this by propagating property info through TypeProcessors so it should work recursively for all "container" types like list, map, map of list, list of map of list, etc.

You can check example in `ObjectAsIdTest` (`testLists`, `testNestedMaps`).

**Comment by panzerj** [comment]
I´ve testet it with my data model and it works as expected - thank you!

Another thing, I would like to mention: you stated in #298 that it is not possible to force Jackson to serialize this class always as object.  .. well I found kind of a workaround to achieve this:

You can place  @JsonIdentityInfo not only on the class, but also on properties or getter functions.

So I removed @JsonIdentityInfo from the class. 
When I want to serialize as Object I just use use the reference without annotations
When I want to serialize as id, I place @JsonIdentityInfo and @JsonIdentityReference(alwaysAsId = true) on the property/method together and get the id in the json.

I don´t know if it is possible that you take this into account?

**Comment by vojtechhabarta** [comment]
I tried to use `@JsonIdentityInfo` on property instead of on class and you are right, using this way it is possible to precisely control whether Jackson serializes particular Java object as JSON object or just ID (and not rely on order in JSON).

So in fea51a276099d496cf00033353ac4a4efc05b382 I added support for it and it should be possible now to produce for some Java property of type `SomeClass` TypeScript type `SomeClass` or `string` (or `SomeClass | string`). Could you please check it?

**Comment by panzerj** [comment]
I testet it with my datamodel and it works as expected! Now you have full control over object/id references! Good work!

**Comment by panzerj** [comment]
Any chances to get this working?
```
public class ValueAssignment<M extends IdData, T> {

    @JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
    @JsonIdentityReference(alwaysAsId = true)
    private M idReference;

    private T assignedValue;
}
```

**Comment by vojtechhabarta** [comment]
This would probably require to generate extra type parameter like `M_ID` (in addition to `M` and `T`) and to fill this parameter on places where `ValueAssignment` would be used. This is relatively complicated so I don't want to implement it. At least not now.

**Comment by vojtechhabarta** [comment]
I released [v2.13.489](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.13.489) with discussed improvements (except for the last one).
